Add Android 14 nonlinear font scaling migration guide (#9425)
flutter/website: src/release/breaking-changes/android-14-nonlinear-text-scaling-migration.md

---
title: Android 14 nonlinear font scaling enabled after v3.14
description: >
  New Android 14 nonlinear font scaling feature is enabled in Flutter after v3.14.
---

## Summary

> Android 14 introduced nonlinear font scaling to 200%. It may change how your app 
> looks when the user changes the accessiblity text scaling in system preferences.

Reviewer: **chore:** Use [semantic breaks](https://sembr.org/). Fix misspelling.
```suggestion
Android 14 introduced nonlinear font scaling up to 200%.
It may change how your app looks when the user changes
the accessibility text scaling in system preferences.
```

## Background

The [Android 14 nonlinear font scaling][] feature prevents excessive accessibility 
font scaling by scaling larger text at a lesser rate when the user increases the 
text scaling value in system preferences.

## Migration guide

> As suggested by the 
> [Android 14 feature overview][Android 14 nonlinear font scaling], perform UI 
> testing with the maximum font size enabled (200%) to ensure that your app 
> applies the font sizes correctly and can accommodate larger font sizes without
> impacting usability.

Reviewer: **suggestion:** Use [semantic breaks](https://sembr.org/). Write in active voice per [GSG](https://developers.google.com/style/voice).
```suggestion
As the [Android 14 feature overview][Android 14 nonlinear font scaling] suggests,
test your UI with the maximum font size enabled (200%).
This should verify that your app can apply the font sizes correctly
and can accommodate larger font sizes without impacting usability.
```

> Consider migrating from `textScaleFactor` to `TextScaler` following this 
> migration guide: [Deprecate `textScaleFactor` in favor of `TextScaler`][], to
> adopt nonlinear font scaling in your app and custom widgets.

Reviewer: **chore:** Condition, then instruction per [GSG](https://developers.google.com/style/sentence-structure). 
Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).

```suggestion
To adopt nonlinear font scaling in your app and custom widgets,
consider migrating from `textScaleFactor` to `TextScaler`.
This migration guide outlines the process:
[Deprecate `textScaleFactor` in favor of `TextScaler`][], .
```

**Temporarily Opting Out**

> If you want to temporarily opt-out of nonlinear text scaling on Android 14 until 
> your app is fully migrated, put a modified `MediaQuery` at the top of your app's 
> widget tree:

Reviewer: **suggestion:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
To opt-out of nonlinear text scaling on Android 14 until you migrate your app,
add a modified `MediaQuery` at the top of your app's widget tree:
```

```dart 
runApp(
  Builder(builder: (context) {
    final mediaQueryData = MediaQuery.of(context);
    final mediaQueryDataWithLinearTextScaling = mediaQueryData
      .copyWith(textScaler: TextScaler.linear(mediaQueryData.textScaler.textScaleFactor));
    return MediaQuery(data: mediaQueryDataWithLinearTextScaling, child: realWidgetTree);
  }),
);
```

> This trick uses the deprecated `textScaleFactor` API and will stop working once
> it's removed from the Flutter API.

Reviewer: **suggestion:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).

```suggestion
This uses the deprecated `textScaleFactor` API.
It will stop working once that API is removed from the Flutter API.
```
